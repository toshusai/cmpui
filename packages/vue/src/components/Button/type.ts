export const ButtonVariant = {
  primary: "primary",
  secondary: "secondary",
} as const;

export type ButtonVariant = (typeof ButtonVariant)[keyof typeof ButtonVariant];

export const ButtonSize = {
  S: "S",
  M: "M",
  L: "L",
} as const;

export type ButtonSize = (typeof ButtonSize)[keyof typeof ButtonSize];

export type ButtonProps = {
  variant: ButtonVariant;
  size: ButtonSize;
  disabled: boolean;
};
