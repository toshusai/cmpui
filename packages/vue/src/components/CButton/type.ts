export const CButtonVariant = {
  primary: "primary",
  secondary: "secondary",
} as const;

export type CButtonVariant =
  (typeof CButtonVariant)[keyof typeof CButtonVariant];

export const CButtonSize = {
  S: "S",
  M: "M",
  L: "L",
} as const;

export type CButtonSize = (typeof CButtonSize)[keyof typeof CButtonSize];

export type CButtonProps = {
  variant?: CButtonVariant;
  size?: CButtonSize;
  disabled?: boolean;
};
