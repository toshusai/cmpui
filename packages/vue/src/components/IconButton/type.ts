export const IconButtonSize = {
  S: "S",
  M: "M",
} as const;

export type IconButtonSize =
  (typeof IconButtonSize)[keyof typeof IconButtonSize];

export type IconButtonProps = {
  bordered?: boolean;
  disabled?: boolean;
  rounded?: boolean;
  selected?: boolean;
  size?: IconButtonSize;
};
