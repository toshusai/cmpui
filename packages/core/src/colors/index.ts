export * from "./blendColors";
export * from "./chooseHigherContrast";
export * from "./componentToHex";
export * from "./const";
export * from "./getContrast";
export * from "./getHighContrast";
export * from "./getLuminance";

export * from "./hexToHsv";

export * from "./hexToRgba";
export * from "./hsvToHex";
export * from "./hsvToRgb";
export * from "./hsvaToRgba";
export * from "./isValidColor";
export * from "./isValidHex";
export * from "./isValidHex6";
export * from "./rgbToCss";
export * from "./rgbToHex";

export * from "./rgbToHsv";
export * from "./rgbaArrayToRgba";
export * from "./rgbaToCss";
export * from "./srgbToLinear";
export * from "./types";

export * from "./useHighContrastColor"