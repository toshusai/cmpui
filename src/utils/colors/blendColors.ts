import { RGB } from "./types/RGB";

export function blendColors(
  foregroundColor: RGB,
  backgroundColor: RGB,
  alpha: number,
): RGB {
  return {
    r: Math.round(foregroundColor.r * alpha + backgroundColor.r * (1 - alpha)),
    g: Math.round(foregroundColor.g * alpha + backgroundColor.g * (1 - alpha)),
    b: Math.round(foregroundColor.b * alpha + backgroundColor.b * (1 - alpha)),
  };
}
