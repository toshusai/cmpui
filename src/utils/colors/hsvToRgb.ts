import { HSV, RGB } from "./types";

/**
 * Reference:
 * - Algorithm from: https://en.wikipedia.org/wiki/HSL_and_HSV#HSV_to_RGB
 */
export function hsvToRgb(hsv: HSV): RGB {
  let r = 0;
  let g = 0;
  let b = 0;

  const h = hsv.h;
  const s = hsv.s;
  const v = hsv.v;

  const i = Math.floor(h / 60);
  const f = h / 60 - i;
  const p = v * (1 - s);
  const q = v * (1 - s * f);
  const t = v * (1 - s * (1 - f));

  switch (i % 6) {
    case 0:
      r = v;
      g = t;
      b = p;
      break;
    case 1:
      r = q;
      g = v;
      b = p;
      break;
    case 2:
      r = p;
      g = v;
      b = t;
      break;
    case 3:
      r = p;
      g = q;
      b = v;
      break;
    case 4:
      r = t;
      g = p;
      b = v;
      break;
    case 5:
      r = v;
      g = p;
      b = q;
      break;
  }

  return {
    r: Math.round(r * 255),
    g: Math.round(g * 255),
    b: Math.round(b * 255),
  };
}
