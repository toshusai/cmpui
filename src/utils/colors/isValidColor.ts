export function isValidColor(color: string) {
  const s = new Option().style;
  s.color = color;
  return s.color !== "";
}
