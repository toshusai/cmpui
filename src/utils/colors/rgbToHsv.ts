import { HSV, RGB } from "./types";

/**
 *
 * References:
 * - https://gist.github.com/mjackson/5311256
 * - https://en.wikipedia.org/wiki/HSL_and_HSV
 * @param r
 * @param g
 * @param b
 * @returns
 */
export function rgbToHsv(rgb: RGB): HSV {
  const red = rgb.r / 255;
  const green = rgb.g / 255;
  const blue = rgb.b / 255;

  const max = Math.max(red, green, blue);
  const min = Math.min(red, green, blue);
  const value = max;

  const chroma = max - min;
  const saturation = value === 0 ? 0 : chroma / value;

  let hue = 0;
  if (chroma === 0) {
    hue = 0;
  } else {
    switch (value) {
      case red:
        hue = (green - blue) / chroma + (green < blue ? 6 : 0);
        break;
      case green:
        hue = (blue - red) / chroma + 2;
        break;
      case blue:
        hue = (red - green) / chroma + 4;
        break;
    }
    hue /= 6;
  }

  return { h: hue * 360, s: saturation, v: value };
}
