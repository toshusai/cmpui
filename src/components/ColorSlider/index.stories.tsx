import type { Meta, StoryObj } from "@storybook/react";
import { useMemo, useState } from "react";
import { ColorSlider } from ".";
import { hsvToRgb } from "../../utils/colors/hsvToRgb";

const meta: Meta<typeof ColorSlider> = {
  title: "Components/Colors/ColorSlider",
  component: ColorSlider,
};

export default meta;
type Story = StoryObj<typeof ColorSlider>;

export const Default: Story = {
  render: function Render() {
    const [value, setValue] = useState(0.5);

    return (
      <ColorSlider
        color={{
          r: (1 - value) * 255,
          g: (1 - value) * 255,
          b: (1 - value) * 255,
          a: value,
        }}
        value={value}
        onChange={setValue}
        bg={
          <div
            style={{
              width: "100%",
              height: "100%",
              backgroundImage:
                "linear-gradient(to right, rgba(0, 0, 0, 0), rgba(0, 0, 0, 1))",
            }}
          />
        }
      />
    );
  },
};

export const Hue: Story = {
  render: function Render() {
    const [value, setValue] = useState(0.5);

    const color = useMemo(() => {
      const { r, g, b } = hsvToRgb({ h: value * 360, s: 1, v: 1 });
      return { r, g, b, a: 1 };
    }, [value]);

    return (
      <ColorSlider
        color={color}
        value={value}
        onChange={setValue}
        bg={
          <div
            style={{
              width: "100%",
              height: "100%",
              backgroundImage:
                "linear-gradient(to right, #f00 0%, #ff0 17%, #0f0 33%, #0ff 50%, #00f 67%, #f0f 83%, #f00 100%)",
            }}
          />
        }
      />
    );
  },
};

export const HeightAndNobSize: Story = {
  render: function Render() {
    const [value, setValue] = useState(0.5);

    return (
      <ColorSlider
        color={{
          r: value * 255,
          g: 0,
          b: 0,
          a: 1,
        }}
        height={32}
        value={value}
        onChange={setValue}
        bg={
          <div
            style={{
              width: "100%",
              height: "100%",
              backgroundImage: "linear-gradient(to right, black, red)",
            }}
          />
        }
      />
    );
  },
};
