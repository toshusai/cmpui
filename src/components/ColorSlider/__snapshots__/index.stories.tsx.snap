// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Colors/ColorSlider Default smoke-test 1`] = `
<div class="cmpui_color-slider__root"
     tabindex="0"
>
  <div style="position: relative; overflow: hidden; cursor: pointer; width: 192px; height: 20px;">
    <div class="alpha-checker-board"
         style="width: 192px; height: 20px;"
    >
      <div style="width: 100%; height: 100%; background-image: linear-gradient(to right, rgba(0, 0, 0, 0), rgb(0, 0, 0));">
      </div>
    </div>
    <div x="96"
         y="10"
         class="cmpui_circle__root alpha-checker-board"
         style="--cmpui-circle-nob-fill: white; --cmpui-circle-nob-radius: 6.666666666666667px; --cmpui-circle-nob-stroke-color: rgb(0,0,0); --cmpui-circle-nob-stroke-width: 2px; border: 2px solid rgb(0, 0, 0); left: 89.3333px; top: 3.33333px; --cmpui-block-size: 4px; pointer-events: none;"
    >
    </div>
    <div x="96"
         y="10"
         class="cmpui_circle__root cmpui_color-slider__nob"
         style="--cmpui-circle-nob-fill: rgba(127.5,127.5,127.5,0.5); --cmpui-circle-nob-radius: 6.666666666666667px; --cmpui-circle-nob-stroke-color: rgb(0,0,0); --cmpui-circle-nob-stroke-width: 2px; border: 2px solid rgb(0, 0, 0); left: 89.3333px; top: 3.33333px;"
    >
    </div>
  </div>
</div>
`;

exports[`Colors/ColorSlider HeightAndNobSize smoke-test 1`] = `
<div class="cmpui_color-slider__root"
     tabindex="0"
>
  <div style="position: relative; overflow: hidden; cursor: pointer; width: 192px; height: 32px;">
    <div class="alpha-checker-board"
         style="width: 192px; height: 32px;"
    >
      <div style="width: 100%; height: 100%; background-image: linear-gradient(to right, black, red);">
      </div>
    </div>
    <div x="96"
         y="16"
         class="cmpui_circle__root alpha-checker-board"
         style="--cmpui-circle-nob-fill: white; --cmpui-circle-nob-radius: 10.666666666666666px; --cmpui-circle-nob-stroke-color: rgb(255,255,255); --cmpui-circle-nob-stroke-width: 2px; border: 2px solid rgb(255, 255, 255); left: 85.3333px; top: 5.33333px; --cmpui-block-size: 4px; pointer-events: none;"
    >
    </div>
    <div x="96"
         y="16"
         class="cmpui_circle__root cmpui_color-slider__nob"
         style="--cmpui-circle-nob-fill: rgba(127.5,0,0,1); --cmpui-circle-nob-radius: 10.666666666666666px; --cmpui-circle-nob-stroke-color: rgb(255,255,255); --cmpui-circle-nob-stroke-width: 2px; border: 2px solid rgb(255, 255, 255); left: 85.3333px; top: 5.33333px;"
    >
    </div>
  </div>
</div>
`;

exports[`Colors/ColorSlider Hue smoke-test 1`] = `
<div class="cmpui_color-slider__root"
     tabindex="0"
>
  <div style="position: relative; overflow: hidden; cursor: pointer; width: 192px; height: 20px;">
    <div class="alpha-checker-board"
         style="width: 192px; height: 20px;"
    >
      <div style="width: 100%; height: 100%; background-image: linear-gradient(to right, rgb(255, 0, 0) 0%, rgb(255, 255, 0) 17%, rgb(0, 255, 0) 33%, rgb(0, 255, 255) 50%, rgb(0, 0, 255) 67%, rgb(255, 0, 255) 83%, rgb(255, 0, 0) 100%);">
      </div>
    </div>
    <div x="96"
         y="10"
         class="cmpui_circle__root alpha-checker-board"
         style="--cmpui-circle-nob-fill: white; --cmpui-circle-nob-radius: 6.666666666666667px; --cmpui-circle-nob-stroke-color: rgb(0,0,0); --cmpui-circle-nob-stroke-width: 2px; border: 2px solid rgb(0, 0, 0); left: 89.3333px; top: 3.33333px; --cmpui-block-size: 4px; pointer-events: none;"
    >
    </div>
    <div x="96"
         y="10"
         class="cmpui_circle__root cmpui_color-slider__nob"
         style="--cmpui-circle-nob-fill: rgba(0,255,255,1); --cmpui-circle-nob-radius: 6.666666666666667px; --cmpui-circle-nob-stroke-color: rgb(0,0,0); --cmpui-circle-nob-stroke-width: 2px; border: 2px solid rgb(0, 0, 0); left: 89.3333px; top: 3.33333px;"
    >
    </div>
  </div>
</div>
`;
